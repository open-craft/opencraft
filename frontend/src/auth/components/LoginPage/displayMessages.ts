const messages = {
  forgotPassword: {
    defaultMessage: 'I forgot my password',
    description: 'Forgot password link.'
  },
  username: {
    defaultMessage: 'Username',
    description: 'Username'
  },
  password: {
    defaultMessage: 'Password',
    description: 'Password'
  },
  login: {
    defaultMessage: 'Login',
    description: 'Text that goes into login button.'
  }
};

export default messages;
