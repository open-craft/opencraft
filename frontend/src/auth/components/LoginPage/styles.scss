@import '~styles/theme';

.login {
  width: 100%;
  padding: 32px 32px;

  .forgot-password {
    text-align: left;
    text-decoration: underline;
    font-family: $chivo-font;
    font-size: 18px;
    letter-spacing: 0.54px;
    color: #1e4c59;
    margin-top: 15px;
  }
}

// FIXME: These are placed in the LoginPage styles, because putting them in `_theme.scss` causes recursive imports.
// Customized alerts
%alert {
  color: #ffffff;
  font-family: $chivo-font;
  font-size: 14px;
  display: flex;
  align-items: center;

  &::before {
    padding-right: 8px;
    font: normal normal normal 24px/1 FontAwesome;
    -webkit-text-stroke: 2px transparent;  // Reduces the "weight" of the icon. Needs specifying a color to work.
    display: inline-flex;
    align-self: center;
  }
}

.alert-success {
  @extend %alert;
  $alert-color: #00a556;
  background-color: $alert-color;

  &::before {
    content: "\f00c";  // fa-check
    -webkit-text-stroke-color: $alert-color;
  }
}

.alert-danger {
  @extend %alert;
  $alert-color: #dd0000;
  background-color: $alert-color;

  &::before {
    content: "\f00d";  // fa-times
    -webkit-text-stroke-color: $alert-color;
  }
}
