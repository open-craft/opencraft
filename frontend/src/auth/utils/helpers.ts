import { V2Api } from 'global/api';

export const checkAuthAndRefreshToken = async () => {
  const refreshToken = localStorage.getItem('token_refresh');
  let response = null;

  if (refreshToken) {
    try {
      response = await V2Api.authRefreshCreate({
        data: { refresh: refreshToken }
      });
    } catch (e) {
      return false;
    }
  }

  if (response) {
    localStorage.setItem('token_access', response.access);
    return true;
  }

  return false;
};
