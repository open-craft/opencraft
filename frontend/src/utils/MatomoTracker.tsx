import React, { ReactNode } from 'react';
import {
  createInstance,
  MatomoProvider,
  useMatomo
} from '@datapunt/matomo-tracker-react';
import {
  MATOMO_ALIAS_DOMAIN,
  MATOMO_BASE_URL,
  MATOMO_MY_DOMAIN,
  MATOMO_SITE_ID
} from '../global/constants';

interface MatomoTrackerProps {
  children: ReactNode;
}

interface MatomoAdditionalConfiguration {
  [propName: string]: any;
}

interface MatomoConfiguration {
  urlBase: string;
  siteId: number;
  configurations?: MatomoAdditionalConfiguration;
}

interface MatomoUserIdTrackerProps {
  userId: string;
}

/*
 * This wrapper component initializes the Matomo tracking by wrapping the children
 * elements, if the tracking is enabled by providing a valid value for the 'MATOMO_BASE_URL'
 * constant. It allows tracking page views, events and the user id, when available.
 *
 * The 'MATOMO_ALIAS_DOMAIN' constant allows the cross-domain linking of another site with
 * this site so that they are treated as the same site in Matomo.
 */
export const MatomoTracker: React.FC<MatomoTrackerProps> = (
  props: MatomoTrackerProps
) => {
  let element = props.children;
  if (MATOMO_BASE_URL) {
    const matomoConfiguration: MatomoConfiguration = {
      urlBase: MATOMO_BASE_URL,
      siteId: parseInt(MATOMO_SITE_ID, 10)
    };
    if (MATOMO_ALIAS_DOMAIN) {
      const myDomain = `*.${MATOMO_MY_DOMAIN}`;
      const aliasDomain = `*.${MATOMO_ALIAS_DOMAIN}`;
      matomoConfiguration.configurations = {
        enableCrossDomainLinking: true,
        setDomains: [myDomain, aliasDomain]
      };
    }
    element = (
      <MatomoProvider value={createInstance(matomoConfiguration)}>
        {props.children}
      </MatomoProvider>
    );
  }
  return <>{element}</>;
};

/**
 * This is a component passes the provided user id to Matomo if Matomo tracking is enabled.
 * The 'pushInstruction' function exposed by the 'useMatomo()' hook is used to do this.
 *
 * This component is needed as all the components in the registration process are classes and hence
 * the 'yseMatomo' hook cannot be directly used in those components.
 */
export const MatomoUserIdTracker: React.FC<MatomoUserIdTrackerProps> = (
  props: MatomoUserIdTrackerProps
) => {
  const { pushInstruction } = useMatomo();
  if (MATOMO_BASE_URL && props.userId) {
    pushInstruction('setUserId', props.userId);
  }
  return null;
};
