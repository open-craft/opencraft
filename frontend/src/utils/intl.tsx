import React from "react";
import { FormattedMessage, MessageDescriptor } from "react-intl";

type MessageMap = { [k: string]: Omit<MessageDescriptor, 'id'> };

interface Props<T, K extends keyof T> {
    id: K;
    messages: T;
    values?: any;
}


export const WrappedMessage = (props: Props<MessageMap, string>) => {
    const {messages, id, ...values} = props;
    return (
        <FormattedMessage  {...messages[id]} id={id} {...values}/>
    );
};
