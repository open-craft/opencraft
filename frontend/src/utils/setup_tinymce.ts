import 'tinymce/tinymce';
import 'tinymce/themes/silver';
import 'tinymce/skins/ui/oxide/skin.min.css';
import 'tinymce/skins/ui/oxide/content.min.css';
import 'tinymce/plugins/advlist';
import 'tinymce/plugins/autolink';
import 'tinymce/plugins/lists';
import 'tinymce/plugins/link';
import 'tinymce/plugins/image';
import 'tinymce/plugins/charmap';
import 'tinymce/plugins/print';
import 'tinymce/plugins/preview';
import 'tinymce/plugins/anchor';
import 'tinymce/plugins/searchreplace';
import 'tinymce/plugins/visualblocks';
import 'tinymce/plugins/code';
import 'tinymce/plugins/fullscreen';
import 'tinymce/plugins/insertdatetime';
import 'tinymce/plugins/media';
import 'tinymce/plugins/table';
import 'tinymce/plugins/paste';
import 'tinymce/plugins/help';
import 'tinymce/plugins/wordcount';
