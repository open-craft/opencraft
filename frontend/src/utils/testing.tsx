import { createMemoryHistory } from 'history';
import * as React from 'react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import * as renderer from 'react-test-renderer';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { createRootReducer } from '../global/reducers';

export const setupComponentForTesting = (
  reactContent: JSX.Element,
  storeContents = {}
) => {
  const middleware = applyMiddleware(thunk);

  const store = createStore(
    createRootReducer(createMemoryHistory()),
    storeContents,
    middleware
  );

  return renderer.create(
    <IntlProvider textComponent={React.Fragment} locale="en">
      <Provider store={store}>
        <MemoryRouter>{reactContent}</MemoryRouter>
      </Provider>
    </IntlProvider>
  );
};
