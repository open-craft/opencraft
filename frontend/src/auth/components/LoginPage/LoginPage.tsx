import * as React from 'react';
import { RootState } from 'global/state';
import { connect } from 'react-redux';
import { ROUTES } from 'global/constants';
import { WrappedMessage } from 'utils/intl';
import { Alert, Button, Spinner, Form } from 'react-bootstrap';
import { ContentPage, TextInputField } from 'ui/components';
import { performLogin } from 'auth/actions';
import messages from './displayMessages';
import './styles.scss';

interface ActionProps {
  performLogin: Function;
}

interface State {
  [key: string]: string | boolean;
  username: string;
  password: string;
}

interface Props extends ActionProps {
  loading: boolean;
  error: string;
}

@connect<{}, ActionProps, {}, Props, RootState>(
  (state: RootState) => state.loginState,
  {
    performLogin
  }
)
export class LoginPage extends React.PureComponent<Props, State> {
  constructor(props: Props, state: State) {
    super(props);

    this.state = {
      username: '',
      password: ''
    };
  }

  private onKeyPress = (event: any) => {
    if (event.key === 'Enter') {
      this.logIn();
    }
  };

  private logIn = () => {
    this.props.performLogin(
      {
        username: this.state.username,
        password: this.state.password
      },
      ROUTES.Console.HOME
    );
  };

  private onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const field = e.target.name;
    const { value } = e.target;

    this.setState({
      [field]: value
    });
  };

  public render() {
    return (
      <ContentPage title="Log in to customize your instance">
        <Form
          className="login"
          onKeyPress={(event: any) => {
            this.onKeyPress(event);
          }}
        >
          <TextInputField
            fieldName="username"
            value={this.state.username}
            onChange={this.onChange}
            messages={messages}
            type="username"
          />
          <TextInputField
            fieldName="password"
            value={this.state.password}
            onChange={this.onChange}
            messages={messages}
            type="password"
          />

          {this.props.error && (
            <Alert variant="danger">{this.props.error}</Alert>
          )}

          <Button
            className="pull-left loading"
            variant="primary"
            size="lg"
            disabled={this.props.loading}
            onClick={() => {
              this.logIn();
            }}
          >
            {this.props.loading && (
              <Spinner animation="border" size="sm" className="spinner" />
            )}
            <WrappedMessage messages={messages} id="login" />
          </Button>

          <p className="forgot-password">
            <a href="/#">
              <WrappedMessage messages={messages} id="forgotPassword" />
            </a>
          </p>
        </Form>
      </ContentPage>
    );
  }
}
