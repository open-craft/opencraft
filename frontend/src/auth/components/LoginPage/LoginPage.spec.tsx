import React from 'react';
import { setupComponentForTesting } from "utils/testing";
import { LoginPage } from './LoginPage';

describe("Login Page", function() {
  it('renders without crashing', () => {
    const tree = setupComponentForTesting(<LoginPage />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders error messages', () => {
    const tree = setupComponentForTesting(
      <LoginPage />,
      {
        loginState: {
          error: "Login error!"
        }
      }
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('disables login button when loading login', () => {
    const tree = setupComponentForTesting(
      <LoginPage />,
      {
        loginState: {
          loading: true
        }
      }
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
