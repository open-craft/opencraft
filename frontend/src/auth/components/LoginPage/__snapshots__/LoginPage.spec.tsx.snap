// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page disables login button when loading login 1`] = `
<div
  className="content-page"
>
  <div
    className="title-container"
  >
    <h1>
      Log in to customize your instance
    </h1>
  </div>
  <div
    className="content-page-container"
  >
    <div
      className="content-page-content row"
    >
      <form
        className="login"
        onKeyPress={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
          >
            Username
          </label>
          <input
            className="form-control"
            name="username"
            onChange={[Function]}
            type="username"
            value=""
          />
          <p />
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
          >
            Password
          </label>
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            type="password"
            value=""
          />
          <p />
        </div>
        <button
          className="pull-left loading btn btn-primary btn-lg"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          <div
            className="spinner spinner-border spinner-border-sm"
          />
          Login
        </button>
        <p
          className="forgot-password"
        >
          <a
            href="/#"
          >
            I forgot my password
          </a>
        </p>
      </form>
    </div>
  </div>
</div>
`;

exports[`Login Page renders error messages 1`] = `
<div
  className="content-page"
>
  <div
    className="title-container"
  >
    <h1>
      Log in to customize your instance
    </h1>
  </div>
  <div
    className="content-page-container"
  >
    <div
      className="content-page-content row"
    >
      <form
        className="login"
        onKeyPress={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
          >
            Username
          </label>
          <input
            className="form-control"
            name="username"
            onChange={[Function]}
            type="username"
            value=""
          />
          <p />
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
          >
            Password
          </label>
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            type="password"
            value=""
          />
          <p />
        </div>
        <div
          className="fade alert alert-danger show"
          role="alert"
        >
          Login error!
        </div>
        <button
          className="pull-left loading btn btn-primary btn-lg"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Login
        </button>
        <p
          className="forgot-password"
        >
          <a
            href="/#"
          >
            I forgot my password
          </a>
        </p>
      </form>
    </div>
  </div>
</div>
`;

exports[`Login Page renders without crashing 1`] = `
<div
  className="content-page"
>
  <div
    className="title-container"
  >
    <h1>
      Log in to customize your instance
    </h1>
  </div>
  <div
    className="content-page-container"
  >
    <div
      className="content-page-content row"
    >
      <form
        className="login"
        onKeyPress={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
          >
            Username
          </label>
          <input
            className="form-control"
            name="username"
            onChange={[Function]}
            type="username"
            value=""
          />
          <p />
        </div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
          >
            Password
          </label>
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            type="password"
            value=""
          />
          <p />
        </div>
        <button
          className="pull-left loading btn btn-primary btn-lg"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Login
        </button>
        <p
          className="forgot-password"
        >
          <a
            href="/#"
          >
            I forgot my password
          </a>
        </p>
      </form>
    </div>
  </div>
</div>
`;
